File: server/fsrs.py
"""
FSRS (Free Spaced Repetition Scheduler) Algorithm Implementation.

Based on the FSRS-4.5 algorithm by Jarrett Ye.
"""

import math
from datetime import datetime, timedelta, timezone

# Rating constants
AGAIN = 1
HARD = 2
GOOD = 3
EASY = 4

# Card states
NEW = "new"
LEARNING = "learning"
REVIEW = "review"
RELEARNING = "relearning"

# Default FSRS parameters (w0..w16)
DEFAULT_WEIGHTS = [
    0.4, 0.6, 2.4, 5.8, 4.93, 0.94, 0.86, 0.01,
    1.49, 0.14, 0.94, 2.18, 0.05, 0.34, 1.26, 0.29, 2.61
]

DECAY = -0.5
FACTOR = 19 / 81  # 0.9^(1/DECAY) - 1


def init_stability(rating: int) -> float:
    """Calculate initial stability for a new card based on first rating.

    S0(G) = w[G-1]
    """
    w = DEFAULT_WEIGHTS
    # rating 1-4 maps to w[0]-w[3]
    return max(w[rating - 1], 0.1)


def init_difficulty(rating: int) -> float:
    """Calculate initial difficulty for a new card based on first rating.

    D0(G) = w[4] - (G-3) * w[5]
    Clamped to [1, 10]
    """
    w = DEFAULT_WEIGHTS
    d = w[4] - (rating - 3) * w[5]
    return min(max(d, 1.0), 10.0)


def next_interval(stability: float) -> int:
    """Calculate the next interval in days from stability.

    I(r, S) = (S / FACTOR) * (r^(1/DECAY) - 1)
    With desired retention r = 0.9:
    I = S * 1  (since (0.9^(1/-0.5) - 1) / FACTOR ≈ 1 when FACTOR = 19/81)

    Simplified: interval = round(stability) with minimum of 1 day.
    """
    # desired retention = 0.9
    # interval = (S / FACTOR) * (R^(1/DECAY) - 1)
    # With R=0.9, DECAY=-0.5: R^(1/DECAY) = 0.9^(-2) = 1/0.81
    # (1/0.81 - 1) = 19/81 = FACTOR
    # So interval = S / FACTOR * FACTOR = S
    interval = round(stability)
    return max(interval, 1)


def retrievability(elapsed_days: float, stability: float) -> float:
    """Calculate the probability of recall after elapsed_days.

    R(t, S) = (1 + FACTOR * t / S)^DECAY
    """
    if stability <= 0:
        return 0.0
    return (1 + FACTOR * elapsed_days / stability) ** DECAY


def next_difficulty(d: float, rating: int) -> float:
    """Calculate next difficulty after a review.

    D'(D, G) = w[7] * D0(3) + (1 - w[7]) * (D - w[6] * (G - 3))
    Clamped to [1, 10]
    """
    w = DEFAULT_WEIGHTS
    d0 = w[4]  # D0(3) = w[4] - 0*w[5] = w[4]
    new_d = w[7] * d0 + (1 - w[7]) * (d - w[6] * (rating - 3))
    return min(max(new_d, 1.0), 10.0)


def next_stability(s: float, d: float, r: float, rating: int) -> float:
    """Calculate next stability after a review.

    For successful recall (rating >= 2):
    S'_r = S * (e^(w[8]) * (11-D) * S^(-w[9]) * (e^(w[10]*(1-R)) - 1) * hard_penalty * easy_bonus + 1)

    For forgetting (rating == 1):
    S'_f = w[11] * D^(-w[12]) * ((S+1)^w[13] - 1) * e^(w[14]*(1-R))
    """
    w = DEFAULT_WEIGHTS

    if rating == AGAIN:
        # Forgetting stability
        new_s = (
            w[11]
            * (d ** (-w[12]))
            * (((s + 1) ** w[13]) - 1)
            * math.exp(w[14] * (1 - r))
        )
        return max(new_s, 0.1)
    else:
        # Successful recall stability
        hard_penalty = w[15] if rating == HARD else 1.0
        easy_bonus = w[16] if rating == EASY else 1.0

        new_s = s * (
            math.exp(w[8])
            * (11 - d)
            * (s ** (-w[9]))
            * (math.exp(w[10] * (1 - r)) - 1)
            * hard_penalty
            * easy_bonus
            + 1
        )
        return max(new_s, 0.1)


def schedule(card_state: dict, rating: int) -> dict:
    """Schedule a card review using FSRS algorithm.

    Args:
        card_state: dict with keys: stability, difficulty, due, last_review, reps, lapses, state
        rating: 1 (Again), 2 (Hard), 3 (Good), 4 (Easy)

    Returns:
        dict with: stability, difficulty, due, interval, state, reps, lapses
    """
    now = datetime.now(timezone.utc)
    state = card_state.get("state", NEW)
    stability = card_state.get("stability", 0.5)
    difficulty = card_state.get("difficulty", 5.0)
    reps = card_state.get("reps", 0)
    lapses = card_state.get("lapses", 0)
    last_review = card_state.get("last_review", "")

    if state == NEW:
        # First review of a new card
        new_stability = init_stability(rating)
        new_difficulty = init_difficulty(rating)
        new_reps = 1
        new_lapses = 0 if rating != AGAIN else 1

        if rating == AGAIN:
            new_state = LEARNING
            interval = 0  # Review again soon (same day)
            due = now + timedelta(minutes=1)
        elif rating == HARD:
            new_state = LEARNING
            interval = 0
            due = now + timedelta(minutes=5)
        elif rating == GOOD:
            new_state = REVIEW
            interval = next_interval(new_stability)
            due = now + timedelta(days=interval)
        else:  # EASY
            new_state = REVIEW
            interval = next_interval(new_stability)
            due = now + timedelta(days=interval)

    elif state in (LEARNING, RELEARNING):
        # In learning/relearning phase
        new_difficulty = next_difficulty(difficulty, rating)
        new_reps = reps + 1

        if rating == AGAIN:
            new_stability = init_stability(AGAIN)
            new_state = state  # Stay in learning/relearning
            new_lapses = lapses + (1 if state == RELEARNING else 0)
            interval = 0
            due = now + timedelta(minutes=5)
        elif rating == HARD:
            new_stability = init_stability(HARD)
            new_state = state
            interval = 0
            due = now + timedelta(minutes=10)
        elif rating == GOOD:
            new_stability = init_stability(GOOD)
            new_state = REVIEW
            new_lapses = lapses
            interval = next_interval(new_stability)
            due = now + timedelta(days=interval)
        else:  # EASY
            new_stability = init_stability(EASY)
            new_state = REVIEW
            new_lapses = lapses
            interval = next_interval(new_stability)
            due = now + timedelta(days=interval)

    else:  # REVIEW state
        # Calculate elapsed days and retrievability
        if last_review:
            try:
                last_dt = datetime.fromisoformat(last_review)
                if last_dt.tzinfo is None:
                    last_dt = last_dt.replace(tzinfo=timezone.utc)
                elapsed = max((now - last_dt).total_seconds() / 86400, 0)
            except (ValueError, TypeError):
                elapsed = 0
        else:
            elapsed = 0

        r = retrievability(elapsed, stability)
        new_stability = next_stability(stability, difficulty, r, rating)
        new_difficulty = next_difficulty(difficulty, rating)
        new_reps = reps + 1

        if rating == AGAIN:
            new_state = RELEARNING
            new_lapses = lapses + 1
            interval = 0
            due = now + timedelta(minutes=5)
        else:
            new_state = REVIEW
            new_lapses = lapses
            interval = next_interval(new_stability)
            due = now + timedelta(days=interval)

    return {
        "stability": round(new_stability, 4),
        "difficulty": round(new_difficulty, 4),
        "due": due.isoformat(),
        "interval": interval,
        "state": new_state,
        "reps": new_reps,
        "lapses": new_lapses,
    }

File: static/js/fsrs.js
// FSRS (Free Spaced Repetition Scheduler) - JavaScript port of server/fsrs.py
// Keeps offline grading identical to server-side scheduling.

export const RATING = { AGAIN: 1, HARD: 2, GOOD: 3, EASY: 4 };
export const STATE = { NEW: 'new', LEARNING: 'learning', REVIEW: 'review', RELEARNING: 'relearning' };

const W = [0.4, 0.6, 2.4, 5.8, 4.93, 0.94, 0.86, 0.01, 1.49, 0.14, 0.94, 2.18, 0.05, 0.34, 1.26, 0.29, 2.61];
const DECAY = -0.5;
const FACTOR = 19 / 81;

function clamp(v, lo, hi) { return Math.min(Math.max(v, lo), hi); }

export function initStability(rating) {
  return Math.max(W[rating - 1], 0.1);
}

export function initDifficulty(rating) {
  const d = W[4] - (rating - 3) * W[5];
  return clamp(d, 1.0, 10.0);
}

export function nextInterval(stability) {
  return Math.max(Math.round(stability), 1);
}

export function retrievability(elapsedDays, stability) {
  if (stability <= 0) return 0.0;
  return Math.pow(1 + FACTOR * elapsedDays / stability, DECAY);
}

export function nextDifficulty(d, rating) {
  const d0 = W[4];
  const newD = W[7] * d0 + (1 - W[7]) * (d - W[6] * (rating - 3));
  return clamp(newD, 1.0, 10.0);
}

export function nextStability(s, d, r, rating) {
  if (rating === RATING.AGAIN) {
    const newS = W[11] * Math.pow(d, -W[12]) * (Math.pow(s + 1, W[13]) - 1) * Math.exp(W[14] * (1 - r));
    return Math.max(newS, 0.1);
  }
  const hardPenalty = rating === RATING.HARD ? W[15] : 1.0;
  const easyBonus = rating === RATING.EASY ? W[16] : 1.0;
  const newS = s * (
    Math.exp(W[8]) * (11 - d) * Math.pow(s, -W[9]) *
    (Math.exp(W[10] * (1 - r)) - 1) * hardPenalty * easyBonus + 1
  );
  return Math.max(newS, 0.1);
}

function addMinutes(date, mins) { return new Date(date.getTime() + mins * 60000); }
function addDays(date, days) { return new Date(date.getTime() + days * 86400000); }

/**
 * Schedule a card review.
 * @param {object} card - {stability, difficulty, due, last_review, reps, lapses, state}
 * @param {number} rating - 1..4
 * @returns {object} - {stability, difficulty, due, interval, state, reps, lapses}
 */
export function schedule(card, rating) {
  const now = new Date();
  const state = card.state || STATE.NEW;
  const stability = card.stability ?? 0.5;
  const difficulty = card.difficulty ?? 5.0;
  const reps = card.reps ?? 0;
  const lapses = card.lapses ?? 0;
  const lastReview = card.last_review || '';

  let newStability, newDifficulty, newReps, newLapses, newState, interval, due;

  if (state === STATE.NEW) {
    newStability = initStability(rating);
    newDifficulty = initDifficulty(rating);
    newReps = 1;
    newLapses = rating === RATING.AGAIN ? 1 : 0;
    if (rating === RATING.AGAIN) {
      newState = STATE.LEARNING; interval = 0; due = addMinutes(now, 1);
    } else if (rating === RATING.HARD) {
      newState = STATE.LEARNING; interval = 0; due = addMinutes(now, 5);
    } else {
      newState = STATE.REVIEW; interval = nextInterval(newStability); due = addDays(now, interval);
    }
  } else if (state === STATE.LEARNING || state === STATE.RELEARNING) {
    newDifficulty = nextDifficulty(difficulty, rating);
    newReps = reps + 1;
    if (rating === RATING.AGAIN) {
      newStability = initStability(RATING.AGAIN);
      newState = state;
      newLapses = lapses + (state === STATE.RELEARNING ? 1 : 0);
      interval = 0; due = addMinutes(now, 5);
    } else if (rating === RATING.HARD) {
      newStability = initStability(RATING.HARD);
      newState = state; newLapses = lapses;
      interval = 0; due = addMinutes(now, 10);
    } else {
      newStability = initStability(rating);
      newState = STATE.REVIEW; newLapses = lapses;
      interval = nextInterval(newStability); due = addDays(now, interval);
    }
  } else { // REVIEW
    let elapsed = 0;
    if (lastReview) {
      const lastDt = new Date(lastReview);
      if (!isNaN(lastDt.getTime())) {
        elapsed = Math.max((now - lastDt) / 86400000, 0);
      }
    }
    const r = retrievability(elapsed, stability);
    newStability = nextStability(stability, difficulty, r, rating);
    newDifficulty = nextDifficulty(difficulty, rating);
    newReps = reps + 1;
    if (rating === RATING.AGAIN) {
      newState = STATE.RELEARNING; newLapses = lapses + 1;
      interval = 0; due = addMinutes(now, 5);
    } else {
      newState = STATE.REVIEW; newLapses = lapses;
      interval = nextInterval(newStability); due = addDays(now, interval);
    }
  }

  return {
    stability: Math.round(newStability * 10000) / 10000,
    difficulty: Math.round(newDifficulty * 10000) / 10000,
    due: due.toISOString(),
    interval,
    state: newState,
    reps: newReps,
    lapses: newLapses,
  };
}

/** Cards due now (due <= now), sorted by due date. */
export function filterDue(cards, now = new Date()) {
  return cards
    .filter((c) => {
      if (!c.due) return true;
      const d = new Date(c.due);
      return !isNaN(d.getTime()) && d <= now;
    })
    .sort((a, b) => new Date(a.due || 0) - new Date(b.due || 0));
}
